Subscription support

w1ll1am23 opened the pull request:
This PR does 2 things. 
1. It adds the ability to pull the subscription key and channel from the initial API query when calling get_devices.
2. It adds support for updating each devices json_state from the data returned form the Pubnub API. For the most part the data returned follows the Wink API.

Three things to note.
1.  Powerstrips needed some tweaking since the Wink API and the Pubnub API treat it as 1 device but we treat it as 2 (1 for each outlet).
2. Spotter humidity is returned as a float in the Pubnub API and not an int like it is in the Wink API. To account for this a check was added to see if humidity was less than 1.0 if so convert to an int. (works unless the humidity is less than 1% which I doubt will happen (don't think it has ever happened any where in the world from a quick Google search))
3. These updates allow for subscription based updates as well as polling updates. Subscription updates will only work when using a token obtained with the Android or iOS client secret and client id. I have a support ticket open with Wink to request subscription keys to be returned in all API requests. It is possible this will never happen of course.

I received the latest update below from Wink on getting subscription details added to all API responses.

> Hi William,
> 
> Victoria here from Wink support. Thanks for reaching out to us. We've identified the issur you're running into and are close to implementing a fix. Once that fix has been pushed out, you'll get the subscription objects you want. I have been assured that this will be very soon. Let me know if you have any other questions or concerns. Thanks!
> 
> Cheers, 
> Victoria

I contacted Wink at the end of last week 4-22-16 . They stated they are still working on this issue and weren't able to provide a date for the fix. I will continue to check the API and get this PR up-to-date once their fix is in place.

Support was added today! This PR should be fully functional. I have been running this since the PR was opened and haven't run into any issues.

bradsk88 commented on src/pywink/devices/standard.py:
```diff
@@ -289,6 +304,9 @@ def _update_state_from_response(self, response_json):
                 outlet['last_reading']['connection'] = power_strip_reading.get('connection')
                 self.json_state = outlet
 
+    def pubnub_update(self, json_response):
```
I'm not sure I see the purpose of all of these `pubnub_update` definitions.  Why wouldn't the library user simply set `device.json_state` directly?

w1ll1am23 replied:
Really, I think it could probably get moved into base and then get overridden. The only one that is required is the powerstrip. This is because the json needs filtered through the _update_state_from_response to get the individual outlet out. Wouldn't want the app to call that or have to handle extracting just the outlets json to set json_state.

Commit:
Retrieve Pubnub details from json if it exists

Commit:
Retrieve Pubnub details from json if it exists

Commit:
Added json example with pubnub details

Commit:
Added json example with pubnub details

Commit:
Fixed powerstrip state update

Commit:
Fixed powerstrip state update

Commit:
Added require_desired_state_fulfilled to overridden method

Commit:
Added require_desired_state_fulfilled to overridden method

Commit:
Added pubnub tests

Commit:
Added pubnub tests

Commit:
Added pubnub json

Commit:
Added pubnub json

Commit:
Get subscription key from first device

Commit:
Get subscription key from first device

bradsk88 commented on src/pywink/devices/standard/__init__.py:
```diff
@@ -129,13 +129,25 @@ def update_state(self, require_desired_state_fulfilled=False):
                 self.json_state = outlet
 
+    def pubnub_update(self, json_response):
+        self._update_state_from_response(json_response)
```
I would prefer to see you "sanitize" the pubnub json response _here_, converting it to a format that is consistent with the output from the Wink API.

What that will likely mean is that some of the devices will need to override `pubnub_update` to handle the inconsistencies.

bradsk88 commented on src/pywink/devices/sensors.py:
```diff
@@ -93,7 +93,13 @@ def humidity_percentage(self):
         :rtype: int
         """
-        return self.last_reading()
+        # The subscription response returns a deciaml not an int
```
See: https://github.com/bradsk88/python-wink/pull/30/files#r62879581

Remove these changes and add an implementation of `pubnub_update` to this class that adapts the pubnub data to match the format of the Wink Data.

bradsk88 commented:
I'd rather not bloat the "pure wink" code in the devices.  Rather, adding `pubnub_update` to the devices, where needed, would be preferred.

bradsk88 commented on CHANGELOG.md:
```diff
@@ -1,5 +1,8 @@
 # Change Log
 
+## 0.7.7
+- Added support for retrieving the Pubnub subscription details
```
Does pubnub require us to provide an endpoint for receiving update events?

How does the pubnub data get to us?

w1ll1am23 replied:
Currently the subscriptions are handled in Home-Assistant. Home-Assistant is making the call to pubnub with a callback to an HA method which then calls self.pubnub_update. This is probably going to change though because HA can't handle the number of connections being made to pubnub.

w1ll1am23 replied:
Turns out that pubnub has a bug that doesn't close previous connections once a new subscriptions is called. This isn't really a problem though because the last connection made to pubnub handles all of the prior subscriptions. So I don't think a change needs made to the current way I am implementing this in Home-Assistant.

bradsk88 commented on src/pywink/devices/standard/__init__.py:
```diff
@@ -129,13 +129,25 @@ def update_state(self, require_desired_state_fulfilled=False):
+        :param response_json: the json obj returned from query
+        :return:
+        """
+        if 'data' in response_json:
```
I would prefer this function no to have to change to support pubnub.  See my other note in this file.

Commit:
Sanitize pubnub json

Commit:
Sanitize pubnub json

Commit:
Added spotter humidity update check

Commit:
Added spotter humidity update check

Commit:
Added pubunub json for spotter

Commit:
Added pubunub json for spotter

bradsk88 commented on src/pywink/api.py:
```diff
@@ -101,6 +101,23 @@ def get_sirens():
+    response = requests.get(arequest_url, headers=API_HEADERS)
+    if response.status_code == 200:
+        response_dict = response.json()
+        first_device = response_dict.get('data')[0]
```
@w1ll1am23 You could do something similar to `get_devices`, which offloads the work of actually processing the API response to `get_devices_from_response_dict`.

You could add `get_subscription_key_from_response_dict` which should be very testable.

bradsk88 commented on src/pywink/api.py:
```diff
@@ -101,6 +101,23 @@ def get_sirens():
+        else:
+            return None
+
+    if response.status_code == 401:
```
This status code checking is duplicated in `get_devices`

You could create a new function called `wink_api_fetch` which handles all of these status codes and either raises or returns the `response_dict`.  Then you could replace `request.get(arequest_url...` with `wink_api_fetch`.

`wink_api_fetch` would be easily testable.  The remainder of `get_subscription_key` and `get_devices` would be so minimal that it probably wouldn't even affect test coverage.

Commit:
Power strip fixes and subscription tests

Commit:
Power strip fixes and subscription tests

Commit:
Fix indent error

Commit:
Fix indent error

bradsk88 commented on src/pywink/devices/sensors.py:
```diff
@@ -95,6 +95,13 @@ def humidity_percentage(self):
         return self.last_reading()
 
+    def pubnub_update(self, json_response):
+        # Pubnub returns the humidity as a decminal
```
"decimal" :)

Commit:
Fixed typo and powerstrip update method

Commit:
Fixed typo and powerstrip update method

bradsk88 committed:
Update README.md

Commit:
Added unit tests for 3 new devices.
GoControl Motion Sensor
GoControl Door Sensor
GE/Jasco Z Wave Light Switch

Stopped duplicating door switches in `get_devices_from_response_dict`

bradsk88 committed:
Merge pull request #41 from python-wink/new-devices

Added unit tests for 3 new devices.

Commit:
Bumping version to 0.7.7

Commit:
Update version

Commit:
Update version

Commit:
Fix rebase error

bradsk88 commented:
Looks good! :dolphin: :+1:
